File: 06 Getting a Little Jumpy/Figure 6-1-2.r
# R code for Figure 6.1.2                 #
# Required packages                       #
# - haven: read .dta files                #
# - data.table: alternative to data.frame #
# - ggplot2: making pretty graphs         #
# - gridExtra: combine graphs             #
library(haven)
library(data.table)
library(ggplot2)
library(gridExtra)

# Download data and unzip the data
# download.file('http://economics.mit.edu/faculty/angrist/data1/mhe/lee', 'Lee2008.zip')
# unzip('Lee2008.zip')

# Load the .dta file as data.table
lee <- data.table(read_dta('Lee2008/individ_final.dta'))

# Subset by non-missing in the outcome and running variable for panel (a)
panel.a <- na.omit(lee[, c("myoutcomenext", "difshare"), with = FALSE])

# Create indicator when crossing the cut-off
panel.a <- panel.a[ , d := (difshare >= 0) * 1.0]

# Predict with local polynomial logit of degree 4
logit   <- glm(formula = myoutcomenext ~ poly(difshare, degree = 4) +
										            poly(difshare, degree = 4) * d,
	             family  = binomial(link = "logit"),
	             data    = panel.a)
panel.a <- panel.a[ , pmyoutcomenext := predict(logit, panel.a, type = "response")]

# Create local average by 0.005 interval of the running variable
breaks  <- round(seq(-1, 1, by = 0.005), 3)
panel.a <- panel.a[ , i005 := as.numeric(as.character(cut(difshare,
	                                                       breaks = breaks,
	                                                       labels = head(breaks, -1),
	                                                       right  = TRUE))), ]

panel.a <- panel.a[ , list(m_next  = mean(myoutcomenext),
	                         mp_next = mean(pmyoutcomenext)),
                   by = i005]

# Plot panel (a)
panel.a <- panel.a[which(panel.a$i005 > -0.251 & panel.a$i005 < 0.251), ]
plot.a  <- ggplot(data = panel.a, aes(x = i005))                       +
           geom_point(aes(y = m_next))                                 +
           geom_line(aes(y = mp_next, group = i005 >= 0))              +
           geom_vline(xintercept = 0, linetype = 'longdash')           +
           xlab('Democratic Vote Share Margin of Victory, Election t') +
           ylab('Probability of Victory, Election t+1')                +
           ggtitle('a')

# Subset the outcome for panel (b)
panel.b <- lee[ , i005 := as.numeric(as.character(cut(difshare,
	                                                   breaks = breaks,
	                                                   labels = head(breaks, -1),
	                                                   right  = TRUE))), ]

panel.b <- panel.b[ , list(m_vic  = mean(mofficeexp, na.rm = TRUE),
	                         mp_vic = mean(mpofficeexp, na.rm = TRUE)),
                   by = i005]

panel.b <- panel.b[which(panel.b$i005 > -0.251 & panel.b$i005 < 0.251), ]
plot.b  <- ggplot(data = panel.b, aes(x = i005))                       +
           geom_point(aes(y = m_vic))                                  +
           geom_line(aes(y = mp_vic, group = i005 >= 0))               +
           geom_vline(xintercept = 0, linetype = 'longdash')           +
           xlab('Democratic Vote Share Margin of Victory, Election t') +
           ylab('No. of Past Victories as of Election t')              +
           ggtitle('b')

lee.p  <- arrangeGrob(plot.a, plot.b)
ggsave(lee.p, file = "Figure 6-1-2-R.png", height = 12, width = 8, dpi = 300)

# End of script

File: 06 Getting a Little Jumpy/Figure 6-1-2.jl
# Load packages
using DataFrames
using Gadfly
using Compose
using GLM

# Download the data and unzip it
# download("http://economics.mit.edu/faculty/angrist/data1/mhe/lee", "Lee2008.zip")
# run(`unzip Lee2008.zip`)

# Read the data
lee = readtable("Lee2008/individ_final.csv")

# Subset by non-missing in the outcome and running variable for panel (a)
panel_a = lee[!isna(lee[:, Symbol("difshare")]) & !isna(lee[:, Symbol("myoutcomenext")]), :]

# Create indicator when crossing the cut-off
panel_a[:d] = (panel_a[:difshare] .>= 0) .* 1.0

# Predict with local polynomial logit of degree 4
panel_a[:difshare2] = panel_a[:difshare].^2
panel_a[:difshare3] = panel_a[:difshare].^3
panel_a[:difshare4] = panel_a[:difshare].^4

logit = glm(myoutcomenext ~ difshare   + difshare2   + difshare3   + difshare4   + d +
                            d*difshare + d*difshare2 + d*difshare3 + d*difshare4,
            panel_a,
            Binomial(),
            LogitLink())
panel_a[:mmyoutcomenext] = predict(logit)

# Create local average by 0.005 interval of the running variable
panel_a[:i005] = cut(panel_a[:difshare], collect(-1:0.005:1))
mean_panel_a   = aggregate(panel_a, :i005, [mean])

# Restrict within bandwidth of +/- 0.251
restriction_a = (mean_panel_a[:difshare_mean] .> -0.251) & (mean_panel_a[:difshare_mean] .< 0.251)
mean_panel_a  = mean_panel_a[restriction_a, :]

# Plot panel (a)
plot_a = plot(layer(x = mean_panel_a[:difshare_mean],
                  y = mean_panel_a[:myoutcomenext_mean],
                  Geom.point),
              layer(x = mean_panel_a[mean_panel_a[:difshare_mean] .< 0, :difshare_mean],
                    y = mean_panel_a[mean_panel_a[:difshare_mean] .< 0, :mmyoutcomenext_mean],
                    Geom.line),
              layer(x = mean_panel_a[mean_panel_a[:difshare_mean] .>= 0, :difshare_mean],
                    y = mean_panel_a[mean_panel_a[:difshare_mean] .>= 0, :mmyoutcomenext_mean],
                    Geom.line),
              layer(xintercept = [0],
                    Geom.vline,
                    Theme(line_style = Gadfly.get_stroke_vector(:dot))),
              Guide.xlabel("Democratic Vote Share Margin of Victory, Election t"),
              Guide.ylabel("Probability of Victory, Election t+1"),
              Guide.title("a"))

# Create local average by 0.005 interval of the running variable
panel_b        = lee[!isna(lee[:, Symbol("difshare")]) & !isna(lee[:, Symbol("mofficeexp")]), :]
panel_b[:i005] = cut(panel_b[:difshare], collect(-1:0.005:1))
mean_panel_b   = aggregate(panel_b, :i005, [mean])

# Restrict within bandwidth of +/- 0.251
restriction_b = (mean_panel_b[:difshare_mean] .> -0.251) & (mean_panel_b[:difshare_mean] .< 0.251)
mean_panel_b  = mean_panel_b[restriction_b, :]

# Plot panel (b)
plot_b = plot(layer(x = mean_panel_b[:difshare_mean],
                  y = mean_panel_b[:mofficeexp_mean],
                  Geom.point),
              layer(x = mean_panel_b[mean_panel_b[:difshare_mean] .< 0, :difshare_mean],
                    y = mean_panel_b[mean_panel_b[:difshare_mean] .< 0, :mpofficeexp_mean],
                    Geom.line),
              layer(x = mean_panel_b[mean_panel_b[:difshare_mean] .>= 0, :difshare_mean],
                    y = mean_panel_b[mean_panel_b[:difshare_mean] .>= 0, :mpofficeexp_mean],
                    Geom.line),
              layer(xintercept = [0],
                    Geom.vline,
                    Theme(line_style = Gadfly.get_stroke_vector(:dot))),             
              Guide.xlabel("Democratic Vote Share Margin of Victory, Election t"),
              Guide.ylabel("No. of Past Victories as of Election t"),
              Guide.title("b"))

# Combine plots
draw(PNG("Figure 6-1-2-Julia.png", 6inch, 8inch), vstack(plot_a, plot_b))

# End of script
